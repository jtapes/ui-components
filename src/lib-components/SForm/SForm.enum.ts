export enum EFormItemTypes {
  INPUT = 'input',
  CHECKBOX = 'checkbox',
  CHECKBOX_GROUP = 'checkbox-group',
  DROPDOWN = 'dropdown',
  RADIO = 'radio',
  RADIO_GROUP = 'radio-group',
  SELECT = 'select',
  PRICERANGE = 'price-range',
  DATEPICKER = 'date-picker',
  AUTOCOMPLETE = 'autocomplete'
}

export enum EFormFieldStatus {
  ACTIVE = 'active',
  DISABLED = 'disabled',
  READONLY = 'readonly',
  LOAD = 'load',
}

export enum EFormFieldComplexLayout {
  GRID_COLUMN = 'grid-column',
  TABLE = 'table',
}

export enum FormValidationMode {
  AGGRESSIVE = 'aggressive',
  LAZY = 'lazy',
  PASSIVE = 'passive',
  EAGER = 'eager'
}
