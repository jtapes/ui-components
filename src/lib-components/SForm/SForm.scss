@import '../../theme/styles/variables';
@import '../../theme/styles/mixins';
@import '../../theme/styles/reset';
@import '../../theme/styles/fonts';
@import '../../theme/styles/global';

.s-form {
  .fields-group__block {
    &--row {
      display: flex;
      flex-wrap: wrap;
    }

    &--grid-column {
      display: grid;
      grid-auto-flow: column;
    }

    &--table {
      display: grid;
      grid-auto-flow: column;
    }
  }
}
