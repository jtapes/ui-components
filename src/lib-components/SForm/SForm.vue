<template>
  <validation-observer ref="form" tag="form" class="s-form">
    <template v-for="(block, block_i) in schema">
      <template v-if="typeof block === 'string'">
        <slot :name="`slot-${block}`" />
      </template>
      <div
        v-else-if="
          isObject(block) &&
            (!block.condition || (block.condition && block.condition()))
        "
        :key="block_i"
        :class="`fields-group mb-40${
          block.groupClass ? ' ' + block.groupClass : ''
        }`"
      >
        <div class="d-f ai-c" :class="{ 'mb-24': block.title }">
          <legend v-if="block.title" class="text bold">
            {{ block.title }}
          </legend>
          <s-hint v-if="block.hint" :content="block.hint" />
        </div>
        <div
          class="fields-group__block"
          :class="
            block.layout === 'row'
              ? `fields-group__block--row`
              : isObject(block.layout)
                ? `fields-group__block--${block.layout.type}`
                : ''
          "
          :style="
            isObject(block.layout) && block.layout.type === 'row' && block.layout.gutter
              ? `
                margin-left: -${block.layout.gutter / 2}px;
                margin-right: -${block.layout.gutter / 2}px;
                `
              : isObject(block.layout) && block.layout.type === 'grid-column'
                ? `
                  grid-template-columns: repeat(${block.layout.columnCount}, 1fr);
                  grid-template-rows: repeat(${Math.ceil(
                    (block.maxLength
                      ? block.fields.slice(0, block.maxLength)
                      : block.fields
                    ).length / block.layout.columnCount
                )}, 1fr);
                  `
                : isObject(block.layout) && block.layout.type === 'table'
                  ? `
                  grid-auto-flow: ${
                    block.layout.flow ? block.layout.flow : 'column'
                  };
                  grid-template-rows: repeat(${
                    block.layout.ratio.split(':')[0]
                  }, auto);
                  grid-template-columns: repeat(${
                    block.layout.ratio.split(':')[1]
                  }, 1fr);
                  ${
                    block.layout.gaps
                      ? (block.layout.gaps.column
                        ? `grid-column-gap: ${block.layout.gaps.column};`
                        : '') +
                        (block.layout.gaps.row
                          ? `grid-row-gap: ${block.layout.gaps.row};`
                          : '')
                      : ''
                  }`
                  : ``
          "
        >
          <template
            v-for="(field, field_i) in block.maxLength
              ? block.fields.slice(0, block.maxLength)
              : block.fields"
          >
            <s-form-item
              v-if="isObject(field)"
              :key="`field-${field_i}`"
              :ref="`formItem_${block_i}_${field_i}`"
              :name="field.valueName"
              :validation-mode="field.validationMode"
              :immediate="field.immediate"
              :mask="field.mask"
              :value="value[field.valueName]"
              :class="`fields-group__field mb-20${
                field.classes ? ' ' + field.classes.join(' ') : ''
              }`"
              :style="getFieldStyles(block, field_i)"
              :component="`s-${field.itemType}`"
              v-bind="field.options ? field.options : []"
              :rules="field.rules
                ? Object.fromEntries(field.rules.map((rule) => [rule.name, rule.params]))
                : {}
              "
              :slots="field.slots ? field.slots : []"
              @change="handleValueChange($event, field)"
            />
            <div
              v-else
              :key="`field-${field_i}`"
              class="fields-group__field mb-20"
              :style="getFieldStyles(block, field_i)"
            >
              <slot :name="`field-${field}`" />
            </div>
          </template>
          <div
            v-if="isBlockError"
            class="mt-8 color-danger"
          >
            {{ block.errorMessage }}
          </div>
        </div>
      </div>
    </template>
  </validation-observer>
</template>

<script lang="ts">
import { Vue, Component, Prop, Model } from 'vue-property-decorator'
import { TFormFieldItem, TFormFieldGroup } from '@/lib-components/SForm/SForm.type'
import SFormItem from '@/lib-components/SFormItem/SFormItem.vue'
import SHint from '@/lib-components/SHint/SHint.vue'
import { isObject } from '@/utils/utils'
import { ValidationObserver } from 'vee-validate'

export enum EFormFieldComplexLayout {
  GRID_COLUMN = 'grid-column',
  TABLE = 'table',
}

export type TFormSchema = Array<string | TFormFieldGroup>
export type FormValidateType = () => Promise<boolean>
export type FormResetValidateType = () => void

@Component({
  components: {
    SFormItem,
    SHint,
    ValidationObserver
  },
})
export default class SForm extends Vue {
  @Model('change')
  value!: any

  @Prop({
    type: Array,
    validator: (val: any[]) =>
      val.every((el) => typeof el === 'string' || isObject(el)),
  })
  schema!: TFormSchema

  $refs!: {
    form: InstanceType<typeof ValidationObserver>
  }

  isBlockError = false
  validate!: FormValidateType
  resetValidation!: FormResetValidateType

  mounted(): void {
    this.$watch(() => this.$refs.form.errors, (val: Record<string, string[]>) => {
      if (val) {
        this.isBlockError = Object.values(val).some((errorsArray) => errorsArray.length)
      }
    }, { deep: true })

    this.validate = this.$refs.form.validate
    this.resetValidation = this.$refs.form.reset
  }

  handleValueChange(value: any, field: TFormFieldItem): void {
    if (typeof field === 'object') {
      const newValue = { ...this.value }
      newValue[field.valueName] = value
      this.$emit('change', newValue)
    }
  }

  getFieldStyles(block: string | TFormFieldGroup, fieldIndex: number): string {
    if (typeof block !== 'string') {
      return isObject(block.layout) && typeof block.layout !== 'string' &&
      block.layout.type === 'row' && block.layout.gutter
        ? `
          padding-left: ${block.layout.gutter / 2}px;
          padding-right: ${block.layout.gutter / 2}px;
        `
        : isObject(block.layout) && typeof block.layout !== 'string' &&
        block.layout.type === EFormFieldComplexLayout.TABLE &&
        block.layout.collapses &&
        Object.keys(block.layout.collapses).includes(String(fieldIndex))
          ? (block.layout.collapses[fieldIndex].column
          ? `grid-column: ${block.layout.collapses[fieldIndex].column}`
          : '') +
          (block.layout.collapses[fieldIndex].row
            ? `grid-row: ${block.layout.collapses[fieldIndex].row}`
            : '')
          : ''
    }
    return ''
  }

  isObject = isObject
}
</script>
