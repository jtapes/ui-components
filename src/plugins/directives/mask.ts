import Vue from 'vue'
import { DirectiveBinding } from 'vue/types/options'

export type TokenType = {
  [key: string]: {
    pattern?: RegExp
    repeat?: boolean
  }
}
export type MaskType = string | { mask: string; tokens: TokenType }

interface MaskedElement extends HTMLInputElement {
  mask?: MaskType
}

const TOKENS = {
  '#': { pattern: /[0-9]/ },
  x: { pattern: /[0-9a-zA-Z]/ },
  X: { pattern: /[0-9а-яА-Яa-zA-Z]/ },
  S: { pattern: /[a-zA-Z]/ },
  E: { pattern: /[0-9a-zA-Z-_@.]/ },
  '*': { repeat: true },
}

function createEvent(name: string) {
  const event = document.createEvent('Event')
  event.initEvent(name, true, true)

  return event
}

function getInputElement(el: HTMLElement) {
  return el instanceof HTMLInputElement ? el : el.querySelector('input')
}

function masking(value: string, mask: string, tokens: TokenType): string {
  let iValue = 0
  let iMask = 0
  let output = ''

  while (iValue < value.length && iMask < mask.length) {
    const token = mask[iMask]
    const maskRegex = tokens[token]
    const currentValue = value[iValue]

    if (maskRegex && maskRegex.pattern) {
      if (maskRegex.pattern.test(currentValue)) {
        output += currentValue
        iMask++
      }

      iValue++
    } else if (maskRegex && maskRegex.repeat) {
      const prevToken = mask[iMask - 1]
      const prevMaskRegex = tokens[prevToken]

      if (prevMaskRegex && prevMaskRegex.pattern) {
        if (!prevMaskRegex.pattern.test(currentValue)) {
          iMask++
        } else {
          iMask--
        }
      }
    } else {
      output += token
      if (currentValue === token) iValue++

      iMask++
    }
  }

  return output
}

function handleValueChange(e: Event) {
  const target = e.target as MaskedElement
  const { mask } = target
  const oldValue = target.value
  if (mask) {
    if (typeof mask === 'string') {
      target.value = masking(oldValue, mask, TOKENS)
    } else {
      target.value = masking(oldValue, mask.mask, mask.tokens)
    }

    if (target.value !== oldValue) {
      target.dispatchEvent(createEvent('input'))
    }
  }
}

function toggleMask(el: HTMLElement, binding: DirectiveBinding) {
  const input = getInputElement(el) as MaskedElement

  if (input && binding.value) {
    input.mask = binding.value
    input.addEventListener('input', handleValueChange)

    if (input.value) {
      input.dispatchEvent(createEvent('input'))
    }
  }
}

export default () => {
  Vue.directive('mask', {
    update(el, binding) {
      if (binding.value && binding.value !== binding.oldValue) {
        Vue.nextTick(() => {
          toggleMask(el, binding)
        })
      }
    },

    bind(el, binding) {
      if (binding.value) {
        toggleMask(el, binding)
      }
    },

    unbind(el, binding) {
      if (binding.value) {
        const input = getInputElement(el)
        if (input) {
          input.removeEventListener('input', handleValueChange)
        }
      }
    },
  })
}
