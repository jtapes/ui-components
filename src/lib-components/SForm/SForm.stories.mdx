import { config, withDesign } from 'storybook-addon-designs'
import { withTests } from '@storybook/addon-jest'
import {
  Meta,
  Story,
  Description,
  Source,
  Preview,
} from '@storybook/addon-docs'
import SForm from '@/lib-components/SForm/SForm.vue'
import results from '/.jest-test-results.json'
import mask from '@/plugins/directives/mask'
import Vue from 'vue'

<Meta
  title="DESIGN SYSTEM/Atom/Form"
  component={SForm}
  decorators={[withDesign, withTests({ results })]}
  argTypes={{
    schema: {
      type: { name: 'array', required: true },
      defaultValue: [],
      description: 'Схема для генерации формы',
      control: { type: 'array' },
    },
    model: {
      type: { name: 'array', required: true },
      defaultValue: [],
      description: 'Объект, хранящий значение всех инпутов формы',
      control: { type: 'object' },
    },
    value: {
      type: 'boolean',
      defaultValue: false,
      description: 'Проходят ли инпуты в форме валидацию',
      control: 'checkbox',
    },
  }}
  parameters={{
    design: config({
      type: 'figma',
      url:
        'https://www.figma.com/file/42Xh2qI3qvq2CRkc2HJlwA/MVP-3-%2F-Market-%2F-Catalog-%E2%80%A2-Products-Page?node-id=132%3A0',
    }),
    docs: {
      source: {
        type: 'code',
      },
    },
  }}
/>

# Генератор формы

<Description of={SForm} />

export const Template = (args, { argTypes }) => {
  Vue.use(mask)
  return {
    props: Object.keys(argTypes),
    components: { SForm },
    template: `
      <SForm
        v-model="model"
        :schema="schema"
      />
    `,
  }
}

<Preview withToolbar withSource="none">
  <Story
    name="SForm"
    args={{
      schema: [
        {
          name: 'sklad',
          title: 'Склад',
          layout: 'column',
          fields: [
            {
              itemType: 'select',
              name: 'storageField',
              valueName: 'storageValue',
              options: {
                items: [
                  'Череповец (СМ)',
                  'Москва (Толбино)',
                  'Воронеж',
                  'Воронеж',
                  'Воронеж',
                ],
                placeholder: 'Город',
                multiple: true,
                searchable: true,
              },
            },
            {
              itemType: 'input',
              name: 'input',
              valueName: 'inputValue',
              options: {
                placeholder: 'Введите значение',
              },
            },
            {
              itemType: 'autocomplete',
              name: 'autocompleteField',
              valueName: 'autocompleteValue',
              options: {
                items: [
                  'Череповец (СМ)',
                  'Москва (Толбино)',
                  'Воронеж',
                  'Воронеж',
                  'Воронеж',
                ],
                placeholder: 'Поиск',
              },
            },
            {
              itemType: 'checkbox',
              name: 'isInStock',
              valueName: 'isInStockValue',
              options: {
                label: 'Сейчас в наличии',
              },
            },
          ],
        },
        {
          name: 'productType',
          title: 'Вид продукта',
          layout: {
            type: 'table',
            ratio: '3:2',
            flow: 'row',
            gaps: {
              column: '12px',
            },
            collapses: {
              0: {
                column: '1 / 3',
              },
            },
          },
          fields: [
            {
              itemType: 'checkbox',
              name: 'productType',
              valueName: 'productTypeValue',
              rules: [{ name: 'required' }],
              options: {
                label: 'Доменный несортовой шлак',
                multiple: true,
                checkedValue: 'sh',
              },
            },
            {
              itemType: 'checkbox',
              name: 'productType',
              valueName: 'productTypeValue',
              options: {
                label: 'Шлак гранулированный',
                multiple: true,
                checkedValue: 'sh2',
              },
            },
            {
              itemType: 'checkbox',
              name: 'productType',
              valueName: 'productTypeValue',
              options: {
                label: 'Щебёночно-песчаная смесь',
                multiple: true,
                checkedValue: 'sh3',
              },
            },
            {
              itemType: 'checkbox',
              name: 'productType',
              valueName: 'productTypeValue',
              options: {
                label: 'Щебень доменный фракционный',
                multiple: true,
                checkedValue: 'sh4',
              },
            },
            {
              itemType: 'checkbox',
              name: 'productType',
              valueName: 'productTypeValue',
              options: {
                label: 'Щебень уличный',
                multiple: true,
                checkedValue: 'sh5',
              },
            }
          ],
        },
      ],
      model: {},
    }}
  >
    {Template.bind({})}
  </Story>
</Preview>

## Пример

<Source
  language="html"
  code={`
<s-form
  v-model="model"
  :schema="schema"
/>
  `}
/>

<ArgsTable of={SForm} />

## Подробнее
Передаём в проп model пустой объект, дальше из него будем просто брать значения
согласно значениям valueName в описаниях полей в массиве fields.
Также передаём схему, про неё подробнее ниже, а также в v-model кидаем булиновскую переменную,
которая будет нам сигнализировать о валидации.

Также при необходимости можно повесить ref на форму и использовать метод triggerValidate()
для проверки валидации. Обязательно его вешаем на любую кнопку для отправки данных, так как изначально
isValid будет true, потому что для фиксации ошибки необходимо, чтобы пользователь "тронул" инпут.

## Схема для пропа schema

<Canvas withSource="none">
  <div>
    <Source
      language="ts"
      code={`
export type TFormFieldGroup = {
  name?: string // Необязательное поле, для более наглядной структуры
  title?: string // Заголовок, рендерится перед очередным блоком с инпутами
  hint?: string // Подсказка рядом с заголовком (без него не использовать)
  maxLength?: number // Ограничение на максимальное число рендерящихся элементов
  groupClass?: string // Можно повесить класс на конкретную группу
  condition?: () => boolean // Условие отображения
  layout: // Настройки рендера
    | 'column' // Обычный рендер в столбик друг за другом
    | 'row' // inline рендер в строку
    | { // рендер в несколько колонок (display: grid)
        type: EFormFieldComplexLayout.GRID_COLUMN
        columnCount: number // количество этих колонок
      }
    | { // табличная вёрстка (display: grid)
        type: EFormFieldComplexLayout.TABLE
        ratio: string // соотношение 'строки:столбцы'
        flow?: 'row' | 'column' // grid-auto-flow
        gaps?: {
          column?: string // расстояние между колонкам (grid-gap-column)
          row?: string // расстояние между столбцами (grid-gap-row)
        }
        collapses?: { // расширения блоков
          [key: number]: { // номер блока, который расширяется
            [key in 'column' | 'row']?: string // в какую сторону и как происходит расширение (grid-column \\ grid-row)
          }
        }
      }
  fields: TFormFieldItem[] // Описание полей
}
// Поле может быть одного из типов
export type TFormFieldItem = TFormFieldInput | TFormFieldCheckbox | TFormFieldRadio | TFormFieldSelect | TFormFieldPriceRange
// Каждое из полей содержит в себе набор базовых параметров:
export type TFormFieldBase = {
  name?: string
  valueName: string // по этому имени будет заноситься значение в объект переданный в проп model
  condition?: () => boolean // условие отрисовки
  classes?: string[] // классы, которые можно вешать на поле
  style?: string // стиль, который можно применять к полю
}
// Остальные подробности лучше читать непосредственно в SForm.vue
      `}
    />
    Пример:
    <Source
      language="js"
      code={`
schema = [
  {
    name: 'blockName',
    title: 'Склад',
    layout: 'column',
    fields: [
      {
        itemType: EFormItemTypes.SELECT,
        name: 'storageField',
        valueName: 'storageValue',
        options: {
          items: [
            'Череповец (СМ)',
            'Москва (Толбино)',
            'Воронеж',
            'Воронеж',
            'Воронеж',
          ],
          placeholder: 'Город',
          multiple: true,
          searchable: true,
        },
      },
      {
        itemType: EFormItemTypes.CHECKBOX,
        name: 'isInStock',
        valueName: 'isInStockValue',
        options: {
          label: 'Сейчас в наличии',
        },
      },
    ],
  },
  'test', // Имя слота, который будет отрендерен согласно порядку в массиве
  {
    name: 'productType',
    title: 'Вид продукта',
    layout: {
      type: EFormFieldComplexLayout.TABLE,
      ratio: '3:2',
      flow: 'row',
      gaps: {
        column: '12px',
      },
      collapses: {
        0: {
          column: '1 / 3',
        },
      },
    },
    fields: [
      {
        itemType: EFormItemTypes.CHECKBOX,
        name: 'productType',
        valueName: 'productTypeValue',
        rules: [{ name: 'required' }],
        options: {
          label: 'Доменный несортовой шлак',
          multiple: true,
          checkedValue: 'sh',
        },
      },
      {
        itemType: EFormItemTypes.CHECKBOX,
        name: 'productType',
        valueName: 'productTypeValue',
        options: {
          label: 'Шлак гранулированный',
          multiple: true,
          checkedValue: 'sh2',
        },
      },
      {
        itemType: EFormItemTypes.CHECKBOX,
        name: 'productType',
        valueName: 'productTypeValue',
        options: {
          label: 'Щебёночно-песчаная смесь',
          multiple: true,
          checkedValue: 'sh3',
        },
      },
      {
        itemType: EFormItemTypes.CHECKBOX,
        name: 'productType',
        valueName: 'productTypeValue',
        options: {
          label: 'Щебень доменный фракционный',
          multiple: true,
          checkedValue: 'sh4',
        },
      },
      {
        itemType: EFormItemTypes.CHECKBOX,
        name: 'productType',
        valueName: 'productTypeValue',
        options: {
          label: 'Щебень уличный',
          multiple: true,
          checkedValue: 'sh5',
        },
      },
    ],
  },
]
      `}
    />
  </div>
</Canvas>
