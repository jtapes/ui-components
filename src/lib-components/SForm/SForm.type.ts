import { SlotType } from '@/lib-components/SFormItem/SFormItem.type'
import { MaskType } from '@/plugins/directives/mask'
import { EInputState } from '@/lib-components/SInput/SInput.type'
import { SCheckboxValue } from '@/lib-components/SCheckbox/SCheckbox.type'
import { SCheckboxGroupItem } from '@/lib-components/SCheckboxGroup/SCheckboxGroup.type'
import { SRadioValue } from '@/lib-components/SRadio/SRadio.type'
import { SSelectItem, SSelectItemValue } from '@/lib-components/SSelect/SSelect.type'
import {
  SAutocompleteItem,
  SAutocompleteItemValue
} from '@/lib-components/SAutocomplete/SAutocomplete.type'
import {
  EFormItemTypes,
  EFormFieldStatus,
  EFormFieldComplexLayout,
  FormValidationMode
} from '@/lib-components/SForm/SForm.enum'
import type { SRadioGroupItem } from '@/lib-components/SRadioGroup/SRadioGroup.type'

export type FormRulesType = {
  name: string
  params?: Array<string | number>
}

export type TFormFieldBase = {
  name: string
  valueName: string
  itemType: string
  condition?: () => boolean
  status?: EFormFieldStatus
  classes?: string[]
  style?: string
  slots?: SlotType[]
}

export type TFormFieldPriceRange = TFormFieldBase & {
  itemType: EFormItemTypes.PRICERANGE
  options: {
    min: number
    max: number
  }
}

export type TFormFieldDatePicker = TFormFieldBase & {
  itemType: EFormItemTypes.DATEPICKER
  options: {
    format?: string
    placeholder?: string
    state?: string
    range?: boolean
  }
}

export type TFormFieldInput = TFormFieldBase & {
  itemType: EFormItemTypes.INPUT
  mask?: MaskType
  rules?: FormRulesType[]
  validationMode?: FormValidationMode
  immediate?: boolean
  options?: {
    type?: string
    placeholder?: string
    clearable?: boolean
    size?: string
    rounded?: boolean
    state?: EInputState
  }
}

export type TFormFieldCheckbox = TFormFieldBase & {
  itemType: EFormItemTypes.CHECKBOX
  rules?: FormRulesType[]
  validationMode?: FormValidationMode
  immediate?: boolean
  options?: {
    label?: string
    multiple?: boolean
    checkedValue?: SCheckboxValue
  }
}

export type TFormFieldCheckboxGroup = TFormFieldBase & {
  itemType: EFormItemTypes.CHECKBOX_GROUP
  rules?: FormRulesType[]
  validationMode?: FormValidationMode
  immediate?: boolean
  options?: {
    items: SCheckboxGroupItem[]
    columns?: number
    searchable?: boolean
    visibleCount?: number
    hiddenState?: string
  }
}

export type TFormFieldRadio = TFormFieldBase & {
  itemType: EFormItemTypes.RADIO
  rules?: FormRulesType[]
  validationMode?: FormValidationMode
  immediate?: boolean
  options: {
    label?: string
    checkedValue: SRadioValue
  }
}

export type TFormFieldRadioGroup = TFormFieldBase & {
  itemType: EFormItemTypes.RADIO_GROUP
  rules?: FormRulesType[]
  validationMode?: FormValidationMode
  immediate?: boolean
  options: {
    items: SRadioGroupItem[]
    columns?: number
    searchable?: boolean
    visibleCount?: number
    hiddenState?: string
  }
}

export type TFormFieldSelect = TFormFieldBase & {
  itemType: EFormItemTypes.SELECT
  rules?: FormRulesType[]
  validationMode?: FormValidationMode
  immediate?: boolean
  options: {
    placeholder?: string
    items: SSelectItem<SSelectItemValue>[] | number[] | string[]
    itemText?: string
    itemValue?: string
    multiple?: boolean
    rounded?: boolean
    searchable?: boolean
    appendToBody?: boolean
  }
}

export type TFormFieldAutocomplete = TFormFieldBase & {
  itemType: EFormItemTypes.AUTOCOMPLETE
  rules?: FormRulesType[]
  validationMode?: FormValidationMode
  immediate?: boolean
  options: {
    placeholder?: string
    items: SAutocompleteItem<SAutocompleteItemValue>[] | number[] | string[]
    itemText?: string
    itemValue?: string
    rounded?: boolean,
    size?: string,
    loading?: boolean
  }
}

export type TFormFieldItem =
  | string
  | TFormFieldInput
  | TFormFieldCheckbox
  | TFormFieldRadio
  | TFormFieldSelect
  | TFormFieldPriceRange
  | TFormFieldDatePicker
  | TFormFieldCheckboxGroup
  | TFormFieldRadioGroup
  | TFormFieldAutocomplete

export type TFormFieldGroup = {
  name?: string
  title?: string
  hint?: string
  maxLength?: number
  groupClass?: string
  condition?: () => boolean
  errorMessage?: string
  layout:
    | 'row'
    | 'column'
    | {
        type: 'row'
        gutter: number
      }
    | {
        type: EFormFieldComplexLayout.GRID_COLUMN
        columnCount: number
      }
    | {
        type: EFormFieldComplexLayout.TABLE
        ratio: string
        flow?: 'row' | 'column'
        gaps?: {
          column?: string
          row?: string
        }
        collapses?: { [key: number]: { [key in 'column' | 'row']?: string } }
      }
  fields: TFormFieldItem[]
}
