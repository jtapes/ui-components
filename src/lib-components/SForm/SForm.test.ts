import { createLocalVue, mount } from '@vue/test-utils'
import SForm from '@/lib-components/SForm/SForm.vue'
import mask from '@/plugins/directives/mask'

describe('SForm', () => {
  test('Is Vue snapshot', () => {
    const wrapper = mount(SForm, {
      propsData: {
        schema: [],
        model: {},
      },
    })
    expect(wrapper.element).toMatchSnapshot()
  })
  test('Simple schema', () => {
    const localVue = createLocalVue()
    localVue.use(mask)
    const wrapper = mount(
      {
        template: '<s-form v-model="model" :schema="schema" />',
        components: { SForm },
        data: () => ({
          model: {
            storageValue: null,
            isInStockValue: null,
          },
          schema: [
            {
              name: 'sklad',
              title: 'Склад',
              layout: 'column',
              fields: [
                {
                  itemType: 'select',
                  name: 'storageField',
                  valueName: 'storageValue',
                  rules: [{ name: 'required' }],
                  options: {
                    items: ['Череповец (СМ)', 'Москва (Толбино)', 'Воронеж'],
                  },
                },
                {
                  itemType: 'checkbox',
                  name: 'isInStock',
                  valueName: 'isInStockValue',
                  rules: [{ name: 'required' }],
                  options: {
                    label: 'Сейчас в наличии',
                  },
                },
              ],
            },
          ],
        }),
      },
      { localVue }
    )
    expect(wrapper.findAll('.fields-group__block').length).toBe(1)
    expect(wrapper.findAll('.fields-group__field').length).toBe(2)
  })
})
