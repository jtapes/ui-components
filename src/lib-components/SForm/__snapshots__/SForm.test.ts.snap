// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SForm Is Vue snapshot 1`] = `
<form
  class="s-form"
  model="[object Object]"
/>
`;
